src/rules/staticThisRule.ts

```diff
@@ -0,0 +1,127 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as utils from "tsutils";
+import * as ts from "typescript";
+
+import * as Lint from "../index";
+import { codeExamples } from "./code-examples/staticThis.examples";
+
+export class Rule extends Lint.Rules.AbstractRule {
+    /* tslint:disable:object-literal-sort-keys */
+    public static metadata: Lint.IRuleMetadata = {
+        ruleName: "static-this",
+        description: "Ban the use of `this` in static methods.",
+        hasFix: true,
+        options: null,
+        optionsDescription: "",
+        optionExamples: [true],
+        /* tslint:disable:max-line-length */
+        rationale: Lint.Utils.dedent`
+            Static \`this\` usage can be confusing for newcomers.
+            It can also become imprecise when used with extended classes when a static \`this\` of a parent class no longer specifically refers to the parent class.
+        `,
+        /* tslint:enable:max-line-length */
+        type: "functionality",
+        typescriptOnly: false,
+        codeExamples,
+    };
+    /* tslint:enable:object-literal-sort-keys */
+
+    public static FAILURE_STRING = (className?: string) => {
+        let message =
+            "`this` usage in static context could be a potential reason of runtime errors.";
+        if (className !== undefined) {
+            /* tslint:disable-next-line:prefer-template */
+            message = "`this` should be replaced with `" + className + "`.";
+        }
+
+        return message;
+    };
+
+    public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+        return this.applyWithWalker(new StaticThisWalker(sourceFile, this.getOptions()));
+    }
+}
+
+const enum ParentType {
+    None,
+    Class,
+    ClassEntity,
+}
+
+class StaticThisWalker extends Lint.RuleWalker {
+    public walk(sourceFile: ts.SourceFile) {
+        let currentParent: ParentType = ParentType.None;
+        let currentClassName: string | undefined;
+
+        const cb = (node: ts.Node): void => {
+            const originalParent = currentParent;
+            const originalClassName = currentClassName;
+
+            if (utils.isClassLikeDeclaration(node)) {
+                currentParent = ParentType.Class;
+
+                if (node.name !== undefined) {
+                    currentClassName = node.name.text;
+                }
+
+                ts.forEachChild(node, cb);
+
+                currentParent = originalParent;
+                currentClassName = originalClassName;
+                return;
+            }
+
+            switch (node.kind) {
```

Comment:
Can you use `hasOwnThisReference` from `tsutils` to save some typing?

Reply:
I've found alternative solution without `hasOwnThisReference` and switch-case spaghetti. Is that ok?

Reply:
😄 awesome, even better!